[Writing Tools] Hide blinking insertion cursor while text placeholders are present

rr-codes opened the pull request:
#### 77eec0a20396c7532b1f63a835b078e57747504c
<pre>
[Writing Tools] Hide blinking insertion cursor while text placeholders are present
<a href="https://bugs.webkit.org/show_bug.cgi?id=278067">https://bugs.webkit.org/show_bug.cgi?id=278067</a>
<a href="https://rdar.apple.com/132432864">rdar://132432864</a>

Reviewed by Aditya Keerthi.

Hide the caret when inserting the placeholder, and unhide it when removing the placeholder.

Also remove some staging declarations.

* Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::insertTextPlaceholder):
(WebCore::Editor::removeTextPlaceholder):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::focusedOrActiveStateChanged):
(WebCore::FrameSelection::addCaretVisibilitySuppressionReason):
(WebCore::FrameSelection::removeCaretVisibilitySuppressionReason):
(WebCore::FrameSelection::updateCaretVisibility):
(WebCore::FrameSelection::setCaretVisibility): Deleted.
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isCaretVisible):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
* Tools/TestWebKitAPI/Tests/TestWebKitAPI/mac/AppKitSPI.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextPlaceholderTests.mm:
(createWebViewForNSTextPlaceholder):
(TEST(NSTextPlaceholder, InsertTextPlaceholder)):
(TEST(NSTextPlaceholder, InsertAndRemoveTextPlaceholderWithoutIncomingText)):
(TEST(NSTextPlaceholder, InsertAndRemoveTextPlaceholderWithIncomingText)):

Canonical link: <a href="https://commits.webkit.org/282500@main">https://commits.webkit.org/282500@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98552fd514dec9ef9e5c20cd7f0a6184b95d7bea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67297 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13884 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65396 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50970 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9582 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31653 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36266 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12142 "Found 1 new test failure: imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/stretch-along-block-axis-001.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12756 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57805 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68993 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58282 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7254 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58506 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6015 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9566 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38453 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39532 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39275 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (commented):
The change seems fine, but we should add a test.

pxlcoder commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1920,7 +1920,16 @@ ExceptionOr<Ref<DOMRect>> Internals::absoluteCaretBounds()
+
+ExceptionOr<bool> Internals::isCaretVisible()
+{
+    Document* document = contextDocument();
```
`RefPtr`

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
```diff
@@ -30,19 +30,6 @@
 #import <AppKit/NSTextInputContext_Private.h>
 #import <AppKit/NSTextPlaceholder_Private.h>
 
-#if HAVE(NSTEXTPLACEHOLDER_RECTS)
-// Staging for rdar://126696059
-@interface NSTextPlaceholder (staging_126696059)
-@property (nonatomic, readonly) NSArray *rects;
-@end
-
-@protocol NSTextInputClient_Async_staging_126696059
-@optional
-- (void)insertTextPlaceholderWithSize:(CGSize)size completionHandler:(void (^)(NSTextPlaceholder *))completionHandler;
-- (void)removeTextPlaceholder:(NSTextPlaceholder *)placeholder willInsertText:(BOOL)willInsertText completionHandler:(void (^)(void))completionHandler;
-@end
-#endif // HAVE(NSTEXTPLACEHOLDER_RECTS)
-
```
Shouldn't this be moved rather than outright deleted, to support building with the public SDK?

rr-codes replied:
fixed

pxlcoder replied:
I'm not seeing the fix?

rr-codes replied:
it's in AppKitSPI.h

rr-codes replied:
and there is already

```
@interface NSTextPlaceholder : NSObject
@property (nonatomic, readonly) NSArray<NSTextSelectionRect *> *rects;
@end
```

in NSTextInputContextSPI.h

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3670,6 +3670,8 @@ RefPtr<TextPlaceholderElement> Editor::insertTextPlaceholder(const IntSize& size
     if (!placeholder->parentNode())
         return nullptr;
 
+    document->selection().setCaretVisible(false);
```
Thinking more about this, would this get overwritten if `FrameSelection::focusedOrActiveStateChanged` gets called in between insertion and removal?

We may end up needing a list of caret visibility suppression "reasons".

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3691,6 +3693,8 @@ void Editor::removeTextPlaceholder(TextPlaceholderElement& placeholder)
     if (document->selection().isFocusedAndActive() && document->focusedElement() == savedRootEditableElement)
         document->selection().setSelection(VisibleSelection { savedPositionBeforePlaceholder }, FrameSelection::defaultSetSelectionOptions(UserTriggered::Yes));
+
+    document->selection().setCaretVisible(true);
```
Similarly if the selection is not focused/active, this may cause us to incorrectly start showing the caret.

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3691,6 +3696,13 @@ void Editor::removeTextPlaceholder(TextPlaceholderElement& placeholder)
+    auto caretVisibilitySuppressionReasons = document->selection().caretVisibilitySuppressionReasons();
+    caretVisibilitySuppressionReasons.remove(CaretVisibilitySuppressionReason::TextPlaceholderIsShowing);
+
+    auto caretVisibility = !caretVisibilitySuppressionReasons.contains(CaretVisibilitySuppressionReason::IsFocusedOrActive) ? CaretVisibility::Visible : CaretVisibility::Hidden;
```
Ideally you'd have a centralized place that keys off of existing reasons and computes a visibility. You shouldn't need to check for one reason in an unrelated part of the code.

rr-codes replied:
fixed

pxlcoder commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -203,14 +203,22 @@ FrameSelection::FrameSelection(Document* document)
+    } else
+        caretVisibilitySuppressionReasons.add(CaretVisibilitySuppressionReason::IsFocusedOrActive);
+
+    auto visibility = activeAndFocused && !caretVisibilitySuppressionReasons.contains(CaretVisibilitySuppressionReason::TextPlaceholderIsShowing) ? CaretVisibility::Visible : CaretVisibility::Hidden;
```
Same comment above would improve this code.

rr-codes commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2268,6 +2268,7 @@ bool FrameSelection::setSelectedRange(const std::optional<SimpleRange>& range, A
 void FrameSelection::focusedOrActiveStateChanged()
 {
     bool activeAndFocused = isFocusedAndActive();
+    auto visibility = activeAndFocused && !caretVisibilitySuppressionReasons.contains() ? CaretVisibility::Visible : CaretVisibility::Hidden;
```
fix this

pxlcoder commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -54,14 +54,19 @@ class VisiblePosition;
+enum class CaretVisibility : bool { Visible, Hidden };
+
+enum class CaretVisibilitySuppressionReason : uint8_t {
+    IsFocusedOrActive,
```
This seems backwards? If the selection is focused/active, we don't want to suppress the caret right?

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3670,6 +3670,8 @@ RefPtr<TextPlaceholderElement> Editor::insertTextPlaceholder(const IntSize& size
     if (!placeholder->parentNode())
         return nullptr;
 
+    document->selection().setCaretVisibility(CaretVisibility::Hidden, { CaretVisibilitySuppressionReason::TextPlaceholderIsShowing });
```
I was thinking more along the lines of adding two methods to `FrameSelection`, like `addCaretVisibilitySuppressionReason` and `removeCaretVisibilitySuppressionReason`. And then `FrameSelection` decides whether to show/hide the caret.

The current code is still a little confusing to read.

pxlcoder commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -204,13 +204,18 @@ FrameSelection::FrameSelection(Document* document)
         m_selection.setIsDirectional(true);
 
     bool activeAndFocused = isFocusedAndActive();
-    if (activeAndFocused)
-        setSelectionFromNone();
```
Why is this being removed?

rr-codes replied:
fixed

pxlcoder reviewed (commented):
List of changed files in the commit message is out of date.

rr-codes committed:
[Writing Tools] Hide blinking insertion cursor while text placeholders are present
https://bugs.webkit.org/show_bug.cgi?id=278067
rdar://132432864

Reviewed by Aditya Keerthi.

Hide the caret when inserting the placeholder, and unhide it when removing the placeholder.

Also remove some staging declarations.

* Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::insertTextPlaceholder):
(WebCore::Editor::removeTextPlaceholder):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
(WebCore::FrameSelection::focusedOrActiveStateChanged):
(WebCore::FrameSelection::addCaretVisibilitySuppressionReason):
(WebCore::FrameSelection::removeCaretVisibilitySuppressionReason):
(WebCore::FrameSelection::updateCaretVisibility):
(WebCore::FrameSelection::setCaretVisibility): Deleted.
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isCaretVisible):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
* Tools/TestWebKitAPI/Tests/TestWebKitAPI/mac/AppKitSPI.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextPlaceholderTests.mm:
(createWebViewForNSTextPlaceholder):
(TEST(NSTextPlaceholder, InsertTextPlaceholder)):
(TEST(NSTextPlaceholder, InsertAndRemoveTextPlaceholderWithoutIncomingText)):
(TEST(NSTextPlaceholder, InsertAndRemoveTextPlaceholderWithIncomingText)):

Canonical link: https://commits.webkit.org/282500@main
